### PR TITLE
Type check block/loop/if sigs

### PR DESCRIPTION
**Pull request by kripken**
This rewrites the type checking code to use block/if/loop sigs properly, as read in s-expr files.

This passes all the spec tests and is basically done I think. but as mentioned in https://github.com/WebAssembly/binaryen/pull/711#issuecomment-250061569 fails on the binary format tests on round-tripping.

This is based on https://github.com/WebAssembly/binaryen/pull/711, so it should be rebased on master after that lands.

### CODE REVIEW

**Comment by kripken** [comment]
This also

 * Prints out if and loop sigs, which was missing and is now necessary.
 * Removes the DSL in OptimizeInstructions (after these type checking changes, using wasm for the DSL became too annoying, not sure what to do there).

File: src/wasm-binary.h [hunk 1]
```diff
@@ -898,6 +918,18 @@ class WasmBinaryWriter : public Visitor<WasmBinaryWriter, void> {
...
+    if (brTargetArities.count(curr->name)) {
+      if (curr->type == unreachable) {
+        arity = brTargetArities[curr->name];
+      } else {
+        assert((curr->type != none) == brTargetArities[curr->name]);
+      }
+    }
+    // For blocks with type unreachable but whose breaks have arity 1, encode i32 as their
+    // signature so that the decoder knows to pop a value for the breaks' values.
+    o << binaryWasmType(curr->type != unreachable ? curr->type : arity ? i32 : none);
```
Comment (dschuff) [inline src/wasm-binary.h]: Once we parse block signatures, we can just encode `curr->type` here which should never be unreachable. Likewise for ifs and loops.

**Comment by kripken** [comment]
Finalizing might still matter for code generators other than wasm-s-parser. But maybe that is worth refactoring later into something on the side.

**Comment by dschuff** [comment]
So the binary format difference is because in the wast parser you replaced the block finalization with the logic that instead only marks the block as unreachable if no breaks target it and the fallthrough is unreachable. If you do that in the binary parser too then it works correctly (I have that working in my local branch).

I do think the "primary" behavior should be that the type of blocks doesn't get changed by default. It might make sense to have a service for binaryen library users that might not know the block type up front (which I guess may include the optimization passes) but it doesn't need to have the same behavior.

**Comment by dschuff** [comment]
In fact in the meantime we could maybe just replace the behavior of the current block `finalize()` with this behavior?

**Comment by kripken** [comment]
Yeah, as discussed in person that's indeed the issue. I'll continue work towards that, with `finalize(type)` on loop/block/if using a given type (as by the s-parser or binary decoder), while `finalize()` with no args autodetects the type fully (which is useful in the optimizer, as you said).

**Comment by kripken** [comment]
Ok, this is hopefully good to go. Tests pass locally.

**Pull request merged**